src/core/receptor.cairo

```diff
@@ -0,0 +1,309 @@
+                let scaled_quote: Wad = scale_x128_to_wad(quote, quote_token_info.decimals);
+
+                quotes.append(scaled_quote);
+                index += 1;
+            }
+        }
+
+        fn get_twap_duration(self: @ContractState) -> u64 {
+            self.twap_duration.read()
+        }
+
+        fn get_update_frequency(self: @ContractState) -> u64 {
+            self.update_frequency.read()
+        }
+
+        fn set_oracle_extension(ref self: ContractState, oracle_extension: ContractAddress) {
+            self.access_control.assert_has_role(receptor_roles::SET_ORACLE_EXTENSION);
+
+            self.set_oracle_extension_helper(oracle_extension);
+        }
+
+        fn set_quote_tokens(ref self: ContractState, quote_tokens: Span<ContractAddress>) {
+            self.access_control.assert_has_role(receptor_roles::SET_QUOTE_TOKENS);
+
+            self.set_quote_tokens_helper(quote_tokens);
+        }
+
+        fn set_update_frequency(ref self: ContractState, new_frequency: u64) {
+            self.access_control.assert_has_role(receptor_roles::SET_UPDATE_FREQUENCY);
+            assert(
+                LOWER_UPDATE_FREQUENCY_BOUND <= new_frequency && new_frequency <= UPPER_UPDATE_FREQUENCY_BOUND,
+                'REC: Frequency out of bounds'
+            );
+
+            self.set_update_frequency_helper(new_frequency);
+        }
+
+        fn set_twap_duration(ref self: ContractState, twap_duration: u64) {
+            self.access_control.assert_has_role(receptor_roles::SET_TWAP_DURATION);
+
+            self.set_twap_duration_helper(twap_duration);
+        }
+
+        fn update_yin_price(ref self: ContractState) {
+            self.access_control.assert_has_role(receptor_roles::UPDATE_YIN_PRICE);
+            self.update_yin_price_internal();
+        }
+    }
+
+    #[abi(embed_v0)]
+    impl ITaskImpl of ITask<ContractState> {
+        fn probe_task(self: @ContractState) -> bool {
+            let seconds_since_last_update: u64 = get_block_timestamp()
+                - self.last_update_yin_price_call_timestamp.read();
+            self.update_frequency.read() <= seconds_since_last_update
+        }
+
+        fn execute_task(ref self: ContractState) {
+            assert(self.probe_task(), 'REC: Too soon to update price');
+            self.update_yin_price_internal();
+        }
+    }
+
+    //
+    // Internal Receptor functions
+    //
+
+    #[generate_trait]
+    impl ReceptorHelpers of ReceptorHelpersTrait {
+        fn set_oracle_extension_helper(ref self: ContractState, oracle_extension: ContractAddress) {
+            assert(oracle_extension.is_non_zero(), 'REC: Zero address for extension');
+
+            self.oracle_extension.write(IEkuboOracleExtensionDispatcher { contract_address: oracle_extension });
+        }
+
+        // Note that this function does not check for duplicate tokens.
+        fn set_quote_tokens_helper(ref self: ContractState, quote_tokens: Span<ContractAddress>) {
```

Comment:
One more thing - can you please refactor this fn to use `mut quote_tokens` and then a loop with pop_front in the body, like we do in other places ([example](https://github.com/lindy-labs/opus_contracts/blob/main/src/core/seer.cairo#L134))?